### PR TITLE
Add Shaderpack Translations

### PR DESCRIPTION
**Pull request by OverlordsIII**
adds translations for shaderpacks parsed from `ShaderPackDir/lang` and then every direct file under that directory

A continuation of that other pr. The reload shaders keybind will be added afterwards/in another pr

### CODE REVIEW

File: src/main/resources/mixins.iris.json [hunk 1]
```diff
@@ -4,16 +4,17 @@
...
-	"texunits.MixinVertexFormats"
+    "MixinGameRenderer",
+    "MixinImmediateVertexConsumerProvider",
+    "MixinWorldRenderer",
+    "texunits.MixinBufferVertexConsumer",
+    "texunits.MixinGlStateManager",
+    "texunits.MixinLightmapTextureManager",
+    "texunits.MixinOverlayTexture",
+    "texunits.MixinVertexFormats",
+    "MixinTranslationStorage"
```
Comment (OverlordsIII) [inline src/main/resources/mixins.iris.json:15]: Everytime I go into this file, intellij just reformats the whole file, I'm not sure why

**Comment by OverlordsIII** [comment]
my git history messed up and while i was switching branches, stuff from one branch spilled into the other, so if this get's merged i have some other things i would like to pr as well

File: src/main/java/net/coderbot/iris/Iris.java [hunk 2]
```diff
@@ -50,11 +59,11 @@ public void onInitializeClient() {
 		if (!irisConfig.isInternal()) {
 			loadExternalShaderpack(irisConfig.getShaderPackName());
 		}
-
```
Comment (coderbot16) [inline src/main/java/net/coderbot/iris/Iris.java:53]: I don't think this deletion is needed

File: src/main/java/net/coderbot/iris/mixin/MixinTranslationStorage.java [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+        //so we also check if the user is loading a special language, and if the shaderpack has support for that language
+        //if they do, we load that, but if they do not, we load "en_us" instead
+        Map<String, Map<String, String>> languageMap = Iris.getCurrentPack().getLangMap();
+        if (!translations.containsKey(key)) {
+            languageCodes.forEach(code -> {
+                if (languageMap.containsKey(code)) {
+                    if (languageMap.get(code).containsKey(key)) {
+                        cir.setReturnValue(languageMap.get(code).get(key));
+                    }
+                }
```
Comment (coderbot16) [inline src/main/java/net/coderbot/iris/mixin/MixinTranslationStorage.java:41]: `containsKey()` and then `get()` can just be simplified to `get()` with a null check

File: src/main/java/net/coderbot/iris/mixin/MixinTranslationStorage.java [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+public class MixinTranslationStorage {
+
+    //this is needed to keep track of which language code we need to grab our lang files from
+    private static List<String> languageCodes = new ArrayList<>();
+
+    private static final String LOAD = "load(Lnet/minecraft/resource/ResourceManager;Ljava/util/List;)Lnet/minecraft/client/resource/language/TranslationStorage;";
+
+    @Shadow @Final private Map<String, String> translations;
+
+    private boolean tested = false;
```
Comment (coderbot16) [inline src/main/java/net/coderbot/iris/mixin/MixinTranslationStorage.java:27]: This seems to be unused

File: src/main/java/net/coderbot/iris/mixin/MixinTranslationStorage.java [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+
+@Mixin(TranslationStorage.class)
+public class MixinTranslationStorage {
+
+    //this is needed to keep track of which language code we need to grab our lang files from
+    private static List<String> languageCodes = new ArrayList<>();
+
+    private static final String LOAD = "load(Lnet/minecraft/resource/ResourceManager;Ljava/util/List;)Lnet/minecraft/client/resource/language/TranslationStorage;";
+
+    @Shadow @Final private Map<String, String> translations;
```
Comment (coderbot16) [inline src/main/java/net/coderbot/iris/mixin/MixinTranslationStorage.java:25]: I prefer the following style:
```suggestion
    @Shadow
    @Final
    private Map<String, String> translations;
```

File: src/main/resources/mixins.iris.json (hunk 1 above)
Comment (coderbot16) [reply src/main/resources/mixins.iris.json:15]: It looks like the indentation on the file was messed before up so this is probably okay

**Review by coderbot16 — CHANGES_REQUESTED** [review]
A few small changes but this mostly looks good, I also noticed that some stuff spilled over but oh well, I don't care enough to remove it for it to just be added back in a future PR

File: src/main/java/net/coderbot/iris/Iris.java [hunk 6]
```diff
@@ -1,16 +1,24 @@
 package net.coderbot.iris;
 
+import com.google.common.base.Throwables;
```
Comment (coderbot16) [inline src/main/java/net/coderbot/iris/Iris.java:3]: Unused import?

File: src/main/java/net/coderbot/iris/Iris.java [hunk 7]
```diff
@@ -1,16 +1,24 @@
...
 import net.fabricmc.api.EnvType;
 import net.fabricmc.api.Environment;
+import net.fabricmc.fabric.api.client.keybinding.v1.KeyBindingHelper;
 import net.fabricmc.loader.api.FabricLoader;
 
+import net.minecraft.client.MinecraftClient;
+import net.minecraft.client.options.KeyBinding;
+import net.minecraft.client.util.InputUtil;
+import net.minecraft.text.TranslatableText;
+import net.minecraft.util.Formatting;
```
Comment (coderbot16) [inline src/main/java/net/coderbot/iris/Iris.java:17]: Unused imports?

File: src/main/java/net/coderbot/iris/Iris.java [hunk 8]
```diff
@@ -1,16 +1,24 @@
...
 import net.coderbot.iris.config.IrisConfig;
 import net.coderbot.iris.pipeline.ShaderPipeline;
 import net.coderbot.iris.shaderpack.ShaderPack;
 import net.fabricmc.api.ClientModInitializer;
 import net.fabricmc.api.EnvType;
 import net.fabricmc.api.Environment;
+import net.fabricmc.fabric.api.client.keybinding.v1.KeyBindingHelper;
 import net.fabricmc.loader.api.FabricLoader;
 
+import net.minecraft.client.MinecraftClient;
```
Comment (coderbot16) [inline src/main/java/net/coderbot/iris/Iris.java:13]: Unused import?

File: src/main/java/net/coderbot/iris/mixin/MixinTranslationStorage.java [hunk 9]
```diff
@@ -0,0 +1,54 @@
...
+        //so we also check if the user is loading a special language, and if the shaderpack has support for that language
+        //if they do, we load that, but if they do not, we load "en_us" instead
+        Map<String, Map<String, String>> languageMap = Iris.getCurrentPack().getLangMap();
+        if (!translations.containsKey(key)) {
+            languageCodes.forEach(code -> {
+                if (languageMap.get(code) != null) {
+                    if (languageMap.get(code).get(key) != null) {
+                        cir.setReturnValue(languageMap.get(code).get(key));
+                    }
+                }
```
Comment (coderbot16) [inline src/main/java/net/coderbot/iris/mixin/MixinTranslationStorage.java:42]: This is what I was getting at with my previous comment.
```suggestion
                Map<String, String> translations = languageMap.get(code);

                if (translations != null) {
                    String translation = translations.get(key);

                    if (translation != null) {
                        cir.setReturnValue(translation);
                    }
                }
```

**Review by coderbot16 — CHANGES_REQUESTED** [review]
Looks mostly good. I noticed that you're using spaces for indentation, in this project tabs are preferred.

**Review by coderbot16 — APPROVED** [review]
Looks good to me, thanks!

**Pull request merged**